### DISCUSSION TITLE
WooCommerce Customizer settings audit

### DISCUSSION THREAD

**Discussion by Aljullu** [woocommerce/woocommerce#52840] [WooCommerce, Blocks, and Site Editing]
WooCommerce comes with some panels under the _Customizer_ (`/wp-admin/customize.php`) which allow changing some specific settings, those include:

* Adding a Store Notice to the site.
* Changing the main query in the Product Catalog.
* Changing the default Product Images thumbnail size and aspect ratio.
* Modifying some options in the Checkout shortcode.

<img src="https://github.com/user-attachments/assets/26884901-e2ae-49da-9d4b-3502ba0b9007" alt="" width="303" />

While the _Customizer_ is easily accessible from the menus in classic themes, it's hidden in most cases in block themes, meaning that most users don't know how to access it.

This has been a source of confusion, as https://github.com/woocommerce/woocommerce/issues/42547 and [this thread](https://wordpress.org/support/topic/woocommerce-catalog-using-300x300-pixel-images-so-they-are-blurry/#post-18133586) prove. Users might change some settings in the _Customizer_ when in a classic theme, switch to a block theme and not find a way to change those settings again.

## Panels

Below I will list what my understanding is of the current plan for them, and whether the setting has any effect on block themes.

### Store Notice

This panel allows writing a text for a Store Notice that will be rendered at the bottom of the screen.

> [!NOTE] 
> 
> * Current plan: In https://github.com/woocommerce/woocommerce/discussions/48526 it was proposed to migrate classic theme users of the _Store Notice_ to a newly created _Store Notice_ block.
> * Does this setting affect block themes? :heavy_check_mark:  Yes

### Product Catalog

This panel has several options:

* _Shop page display_ and _Category display_: this allows displaying products and categories in the same grid. This is something not currently supported in the Product Catalog block.
* _Default product sorting_: this is not supported by the Product Collection block when _Query Type_ is set to _Default_.
* _Products per row_ and _Rows per page_: both are supported by the Product Collection block.

> [!NOTE] 
> 
> * Current plan: AFAIK we don't have any short-term plans to implement the possibility to display categories inside the Product Collection block.
> * Does this setting affect block themes?
>     * :heavy_check_mark:  Yes (if the theme uses the _Classic Template_ block)
>     * :stop_sign:  Partially: sorting options work and the _Products per row_ and _Rows per page_ options impact the total number of products displayed, but not the layout (if the theme uses the _Product Catalog_ block)

### Product Images

This allows defining the width of the thumbnails generated by WooCommerce as well as their aspect ratio. We currently don't have anything similar in blocks.

> [!NOTE] 
> 
> * Current plan: In https://github.com/woocommerce/woocommerce/discussions/49343 I made a proposal to move away from WooCommerce-generated thumbnails.
> * Does this setting affect block themes? :heavy_check_mark:  Yes

### Checkout

This panel has several options for the Checkout shortcode.

> [!NOTE] 
> 
> * Current plan: AFAIK there is no plan to move these settings to the block editor. Even though, it would be possible given that we have the _Classic Checkout_ block, which is a wrapper of the _Checkout_ shortcode.
> * Does this setting affect block themes? :heavy_check_mark:  Yes (even though, I would expect most block themes to use the _Checkout_ block)

## A path forward

After a discussion with @gigitux, we would like to propose a path forward:

1. When the user switches from a classic theme to a block theme, we will 'reset' all _Customizer_ options to their defaults. This will help avoid confusion when users switch from a classic theme to a block theme and don't see the possibility to edit those options. Given that this is a quick-fix, we should ship it without waiting for step 2 to be completed.
    * Note: that wouldn't affect existing sites that migrated from a classic theme to a block theme in the past, to avoid breaking live stores.
2. If we wanted to re-implementing the same features from the _Customizer_ into blocks, that would involve:
   1. Creating a Store Notice block.
   2. Updating the Product Collection block to allow displaying product categories in the grid and enabling the _Order by_ option when _Query Type_ is set to _Default_.
   3. Work on proposal 2 of #49343 and/or work on a way so users can modify product image settings.
   4. Add _Checkout_'s Customizer settings to the _Classic Checkout_ block.

## Questions to answer

1. Do you agree to reset the _Customizer_ options when switching from a classic theme to a block theme? Keep in mind that when switching between classic themes, most _Customizer_ settings are persisted, so that would be a different behavior. However, it could be considered expected given that block themes are an entirely different paradigm.
    * A subquestion is: should the options be restored when switching back to a classic theme?
    * Another subquestion that needs to be answered: do we want to do it for all settings/panels? Ie, are we ok with block themes using the _Checkout_ shortcode to ignore the settings from the _Customizer_?
2. What do you think about the tasks to re-implement the _Customizer_ settings into blocks? Should we work on all of them? Now it's a good opportunity to break from previous features we consider low-priority, so we might opt towards not implementing some if we think they are not worth it.

_(Thanks @gigitux for reviewing this post and making suggestions)_

**Comment by mikejolley** [comment]
> Do you agree to reset the Customizer options when switching from a classic theme to a block theme? Keep in mind that when switching between classic themes, most Customizer settings are persisted, so that would be a different behavior. However, it could be considered expected given that block themes are an entirely different paradigm.

We're looking at moving the checkout block attributes for field visibility to globally accessible settings. Because core already uses the settings from the customizer, we're looking into reusing those settings for this purpose.

https://github.com/woocommerce/woocommerce/pull/52784

Therefore in the case of checkout, resetting the fields would not be ideal. They can persist between customizer and block editor.

**Reply by Aljullu** [reply]
Oh, interesting. Would a block theme store using the block shortcode be able to modify those settings in some way that doesn't involve going to the _Customizer_?

**Comment by ralucaStan** [comment]
Also, there are more Checkout Customizer fields outside of the Company name, Address line 2 field & Phone visibility & requirement settings.


<img width="290" alt="Screenshot 2024-11-14 at 19 59 57" src="https://github.com/user-attachments/assets/d76a5546-8704-4515-aa46-1f5aea037d90">

**The fields below:** 

> Highlight required fields with an asterisk

This no longer applies to the Checkout block and there was no report about it, if we think at sites that use the shortcode and change themes. I think we can easily ignore it.

> Terms and conditions page

This is a shortcut for the field on `wp-admin/admin.php?page=wc-settings&tab=advanced`. Both the checkout block and the checkout shortcode use this as the source of truth. No need to reset it.

> Privacy policy & Terms and conditions

These fields are not used by the checkout block at all, but they are used by the checkout shortcode. I don't think it will make a difference when switching just the type of theme. So I think it can be ignored

**Reply by Aljullu** [reply]
Thanks for taking a look at each setting in detail, @ralucaStan!

**Comment by maxlaf** [comment]
Thank you Albert and Luigi for the incredibly thorough proposal!

>Current plan: AFAIK we don't have any short-term plans to implement the possibility to display categories inside the Product Collection block.

Not really related to this proposal, but this is a gap that I've seen come up more often these days and have seen several pings about in the last month or two alone. There isn't a good 1:1 replacement in PC for the Display Product Categories setting in the Customizer. There are some unintuitive and hacky workarounds, but not ideal.

I can grab some Slack links or example tickets if needed, but I do think we need to explore some implementation of this in PC.

>When the user switches from a classic theme to a block theme, we will 'reset' all Customizer options to their defaults.

I believe the default for the Store Notice is enabled and it contains text about a demo store. Are we changing this default?


>Do you agree to reset the Customizer options when switching from a classic theme to a block theme?

I'll be thinking about this further, but my initial sense is, I agree. But, I think the settings should be restored as you mentioned here


>A subquestion is: should the options be restored when switching back to a classic theme?

Users will often ask "if I change themes, do I lose my settings?" With the exception of poorly coded themes that I've seen destroy sites, the answer is "No". If we reset customizer settings, it won't be a "No" anymore, it'll be a "Maybe" and could cause concern.


>What do you think about the tasks to re-implement the Customizer settings into blocks? Should we work on all of them?

I think providing PC with a way to display Product Categories would be the most impactful. Others, like a Store Notice block, can be easily recreated with basic GB blocks. 

If we made a Store Notice block, I think it would tie Block Themes to Customizer settings for Classic Themes in a way that could be harder to break in the future. In other words, a Store Notice block seems like a step back to me and it's main use would be to allow the usage of a legacy customizer setting that's not needed anymore.

**Reply by Aljullu** [reply]
> Not really related to this proposal, but this is a gap that I've seen come up more often these days and have seen several pings about in the last month or two alone. There isn't a good 1:1 replacement in PC for the Display Product Categories setting in the Customizer. There are some unintuitive and hacky workarounds, but not ideal.
> 
> I can grab some Slack links or example tickets if needed, but I do think we need to explore some implementation of this in PC.

Completely agree. :100: Sorry if I wasn't clear. I didn't mean I want it not to be implemented (actually, it's the opposite, I would love to see this!), but mostly that we don't have this planned in the short-term.

> I believe the default for the Store Notice is enabled and it contains text about a demo store. Are we changing this default?

I just tested and no, the Store Notice is not enabled by default in my case. I tried creating a new site (with TT5 by default), then switched to Twenty Nineteen and then to Storefront and in all themes the Store Notice wasn't visible.

> I'll be thinking about this further, but my initial sense is, I agree. But, I think the settings should be restored as you mentioned here
> 
> Users will often ask "if I change themes, do I lose my settings?" With the exception of poorly coded themes that I've seen destroy sites, the answer is "No". If we reset customizer settings, it won't be a "No" anymore, it'll be a "Maybe" and could cause concern.

Good points, it makes sense restoring the settings when switching back to a classic theme.

> I think providing PC with a way to display Product Categories would be the most impactful. Others, like a Store Notice block, can be easily recreated with basic GB blocks.
> 
> If we made a Store Notice block, I think it would tie Block Themes to Customizer settings for Classic Themes in a way that could be harder to break in the future. In other words, a Store Notice block seems like a step back to me and it's main use would be to allow the usage of a legacy customizer setting that's not needed anymore.

:+1:

**Comment by pmcpinto** [comment]
IIRC we disable the Checkout menu item when the Checkout block is enabled: https://github.com/woocommerce/woocommerce/issues/44120

@ralucaStan am I missing something?

**Reply by Aljullu** [reply]
> IIRC we disable the Checkout menu item when the Checkout block is enabled: https://github.com/woocommerce/woocommerce/issues/44120

Oh, you're right, TIL. However, I think there is still the case of a classic theme store using the shortcode Checkout switching to a block theme. In that case, the shortcode Checkout is still the one being used but there is no way to access its settings, right?

**Comment by IanMisnerThough** [comment]
The new Site Visibility options might be a good place to consider moving the Store Notice concept into to keep it consistent across all theme types. It's a pretty wide-reaching "show this notice everywhere" concept that matches the context of that feature.

**Reply by Aljullu** [reply]
I didn't consider this idea, thanks @IanMisnerThough! I agree the Site Visibility options might be a good place for the Store Notice options. However, as @maxlaf [said above](https://github.com/woocommerce/woocommerce/discussions/52840#discussioncomment-11339195), I lean towards not having a Store Notice option in block themes and simply let blocks handle that, but it's not a strong opinion.

**Comment by Aljullu** [comment]
Heads-up that I updated the post above, as I just realized that _Default product sorting_ is not supported by the Product Collection block when _Query Type_ is set to _Default_. So this is something that will need to be implemented in the Product Collection block if we want to achieve feature parity.

**Comment by Aljullu** [comment]
I went ahead and created individual issues for each of the tasks, so it's easier to pick them individually. I based them on my original post and all the responses/suggestions it got, but feel free to comment if you don't agree with the proposed solutions.

* Store Notice ( :heavy_check_mark: completed in WooCommerce 9.7): https://github.com/woocommerce/woocommerce/issues/54237
* Product Catalog:
  * Shop page display and Category display ( :stop_sign:  not started): https://github.com/woocommerce/woocommerce/issues/54238
  * Default product sorting ( :heavy_check_mark: completed in WooCommerce 9.8): https://github.com/woocommerce/woocommerce/pull/54166
  * Products per row and Rows per page ( :stop_sign:  not started): https://github.com/woocommerce/woocommerce/issues/54240
* Product Images ( :stop_sign:  not started): https://github.com/woocommerce/woocommerce/issues/54241
* Checkout ( :heavy_check_mark: completed in WooCommerce 9.6): https://github.com/woocommerce/woocommerce/pull/52784
